query-algebrizer/src/types.rs

```diff
@@ -461,4 +464,140 @@ impl Debug for EmptyBecause {
+        let mut o = EnumSet::new();
+        o.insert(x);
+        o.insert(y);
+        o
+    }
+
+    /// Return a clone of `self` with `y` added.
+    fn with(&self, y: T) -> EnumSet<T> {
+        let mut o = self.clone();
+        o.insert(y);
+        o
+    }
+}
+
+impl ValueTypeSet {
+    /// Returns a set containing all the types in this set and `other`.
+    pub fn union(&self, other: &ValueTypeSet) -> ValueTypeSet {
+        match self {
+            &ValueTypeSet::None => other.clone(),
+            &ValueTypeSet::Any => ValueTypeSet::Any,
+            &ValueTypeSet::One(t) if other.contains(t) => other.clone(),
+            &ValueTypeSet::One(t) => {
+                match other {
+                    &ValueTypeSet::None => self.clone(),
+                    &ValueTypeSet::Any => ValueTypeSet::Any,
+                    &ValueTypeSet::One(o) if t == o => self.clone(),
+                    &ValueTypeSet::One(o) => ValueTypeSet::Many(EnumSet::both(o, t)),
+                    &ValueTypeSet::Many(os) => ValueTypeSet::Many(os.with(t)),
+                }
+            },
+            &ValueTypeSet::Many(ts) => {
+                match other {
+                    &ValueTypeSet::None => self.clone(),
+                    &ValueTypeSet::Any => ValueTypeSet::Any,
+                    &ValueTypeSet::One(o) if ts.contains(&o) => self.clone(),
+                    &ValueTypeSet::One(o) => ValueTypeSet::Many(ts.with(o)),
+                    &ValueTypeSet::Many(os) => ValueTypeSet::Many(os.union(ts)),
+                }
+            },
+        }
+    }
+
+    pub fn intersection(&self, other: &ValueTypeSet) -> ValueTypeSet {
+        match (self, other) {
+            (&ValueTypeSet::None, _) => ValueTypeSet::None,
+            (_, &ValueTypeSet::None) => ValueTypeSet::None,
+            (s, &ValueTypeSet::Any) => s.clone(),
+            (&ValueTypeSet::Any, s) => s.clone(),
+            (&ValueTypeSet::One(t), s) => if s.contains(t) { ValueTypeSet::One(t) } else { ValueTypeSet::None },
+            (s, &ValueTypeSet::One(t)) => if s.contains(t) { ValueTypeSet::One(t) } else { ValueTypeSet::None },
+            (&ValueTypeSet::Many(ref left), &ValueTypeSet::Many(ref right)) => {
+                let result: EnumSet<ValueType> = left.intersection(right.clone());
+                match result.len() {
+                    0 => ValueTypeSet::None,
+                    1 => ValueTypeSet::One(result.into_iter().next().unwrap()),
+                    _ => ValueTypeSet::Many(result),
+                }
+            },
+        }
+    }
+
+    /// Return an arbitrary type that's part of this set.
+    pub fn exemplar(&self) -> Option<ValueType> {
+        match self {
+            &ValueTypeSet::None => None,
+            &ValueTypeSet::Any => Some(ValueType::Ref),
+            &ValueTypeSet::One(t) => Some(t),
+            &ValueTypeSet::Many(ref s) => s.iter().next(),
+        }
+    }
+
+    pub fn contains(&self, vt: ValueType) -> bool {
```

Comment:
The `contains` call for collections usually does, but `ValueType` is `Copy` and probably only 32 or 64 bits wide, so it's cheaper to pass by copy than by reference.